pkg/services/authn/authnimpl/sync/rbac_sync.go

```diff
@@ -34,19 +35,57 @@ func (s *RBACSync) SyncPermissionsHook(ctx context.Context, ident *authn.Identit
 		return nil
 	}
 
-	permissions, err := s.ac.GetUserPermissions(ctx, ident, accesscontrol.Options{ReloadCache: false})
+	// Populate permissions from roles
+	permissions, err := s.fetchPermissions(ctx, ident)
 	if err != nil {
-		s.log.FromContext(ctx).Error("Failed to fetch permissions from db", "error", err, "id", ident.ID)
-		return errSyncPermissionsForbidden
+		return err
 	}
 
 	if ident.Permissions == nil {
-		ident.Permissions = make(map[int64]map[string][]string)
+		ident.Permissions = make(map[int64]map[string][]string, len(permissions))
+	}
+	grouped := accesscontrol.GroupScopesByAction(permissions)
+
+	// Restrict access to the list of actions
+	actionsLookup := ident.ClientParams.FetchPermissionsParams.ActionsLookup
+	if len(actionsLookup) > 0 {
+		filtered := make(map[string][]string, len(ident.ClientParams.FetchPermissionsParams.ActionsLookup))
+		for _, action := range ident.ClientParams.FetchPermissionsParams.ActionsLookup {
+			if scopes, ok := grouped[action]; ok {
+				filtered[action] = scopes
+			}
+		}
+		grouped = filtered
 	}
-	ident.Permissions[ident.OrgID] = accesscontrol.GroupScopesByAction(permissions)
+
+	ident.Permissions[ident.OrgID] = grouped
 	return nil
 }
 
+func (s *RBACSync) fetchPermissions(ctx context.Context, ident *authn.Identity) ([]accesscontrol.Permission, error) {
+	permissions := make([]accesscontrol.Permission, 0, 8)
+	roles := ident.ClientParams.FetchPermissionsParams.Roles
+	if len(roles) > 0 {
+		for _, role := range roles {
+			roles, err := s.ac.GetRoleByName(ctx, ident.GetOrgID(), role)
```

Comment:
nit: could rename `roles` to `roleDTO` since it's only one role returned. Plus it's shadowing the `roles` variable above.